sdk/storage/Azure.Storage.Queues/tests/QueueClientTests.cs

```diff
@@ -754,6 +757,58 @@ public async Task DecodesPeekedMessage()
             Assert.AreEqual(messageText, receivedMessage.MessageText);
         }
 
+        [Test]
+        public async Task FailsOnInvalidPeekedMessageIfNoHandlerIsProvided()
+        {
+            // Arrange
+            await using DisposingQueue test = await GetTestQueueAsync();
+            var encodingClient = GetEncodingClient(test.Queue.Name, QueueMessageEncoding.Base64);
+            var nonEncodedContent = "test_content";
+
+            await test.Queue.SendMessageAsync(nonEncodedContent);
+
+            // Act
+            await TestHelper.AssertExpectedExceptionAsync<FormatException>(
+                encodingClient.PeekMessagesAsync(),
+                e =>
+                {
+                    StringAssert.Contains("The input is not a valid Base-64 string", e.Message);
+                });
+        }
+
+        [Test]
+        public async Task CanHandleInvalidPeekedMessageAndReturnValid()
+        {
+            // Arrange
+            await using DisposingQueue test = await GetTestQueueAsync();
+            object badMessage = null;
+            var encodingClient = GetEncodingClient(
+                test.Queue.Name,
+                QueueMessageEncoding.Base64,
+                arg =>
+                {
+                    badMessage = arg.Message;
+                    return Task.CompletedTask;
+                });
+            var nonEncodedContent = "test_content";
+
+            await test.Queue.SendMessageAsync(nonEncodedContent);
+            await encodingClient.SendMessageAsync(nonEncodedContent);
+
+            // Act
+            PeekedMessage[] peekedMessages = await encodingClient.PeekMessagesAsync(10);
+
+            // Assert
+            Assert.AreEqual(1, peekedMessages.Count());
+            await WaitForCondition(() => badMessage != null);
```

Comment:
I would use TaskCompletionSource to notify the event firing. It makes tests faster and more deterministic.

Reply:
that won't be necessary with new design.